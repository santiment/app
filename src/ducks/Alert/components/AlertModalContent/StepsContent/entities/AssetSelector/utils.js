export function hasAssetById({ id, listItems }) {
  if (!id || !listItems) return
  return listItems.some(({ id: projectId }) => projectId === id)
}
