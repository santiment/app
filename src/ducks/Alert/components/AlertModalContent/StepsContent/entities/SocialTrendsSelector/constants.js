export const SOCIAL_TRENDS_OPTIONS = [
  {
    label: 'Trending assets',
    type: 'slug',
  },
  {
    label: 'Trending words',
    type: 'word',
  },
  {
    label: 'Watchlist',
    type: 'watchlist_id',
  },
]
