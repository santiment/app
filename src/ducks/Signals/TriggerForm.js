import React, { useEffect, useState } from 'react'
import { Link } from 'react-router-dom'
import PropTypes from 'prop-types'
import isEqual from 'lodash.isequal'
import cx from 'classnames'
import { compose } from 'recompose'
import { graphql } from 'react-apollo'
import { Formik, Form } from 'formik'
import { connect } from 'react-redux'
import {
  PanelWithHeader as Panel,
  Button,
  Message
} from '@santiment-network/ui'
import { selectIsTelegramConnected } from './../../pages/UserSelectors'
import { allProjectsForSearchGQL } from './../../pages/Projects/allProjectsGQL'
import { fetchHistorySignalPoints } from './actions'
import FormikInput from './../../components/formik-santiment-ui/FormikInput'
import FormikSelect from './../../components/formik-santiment-ui/FormikSelect'
import FormikSelector from './../../components/formik-santiment-ui/FormikSelector'
import FormikCheckboxes from './../../components/formik-santiment-ui/FormikCheckboxes'
import FormikToggle from './../../components/formik-santiment-ui/FormikToggle'
import FormikEffect from './FormikEffect'
import SignalPreview from './SignalPreview'
import styles from './TriggerForm.module.scss'

const METRICS = [
  { label: 'Price', value: 'price' },
  // { label: 'Trending Words', value: 'trendingWords' },
  { label: 'Daily Active Addresses', value: 'daily_active_addresses' },
  { label: 'Price/volume difference', value: 'price_volume_difference' }
]

const OPERATIONS = {
  price_absolute_change: [
    'above',
    'below',
    'inside_channel',
    'outside_channel'
  ],
  price_percent_change: ['percent_up', 'percent_down']
}

const TYPES = {
  price: [
    {
      label: 'Price changing',
      options: [
        {
          value: 'above',
          label: 'More than'
        },
        {
          value: 'below',
          label: 'Less than'
        },
        {
          value: 'inside_channel',
          label: 'Entering channel'
        },
        {
          value: 'outside_channel',
          label: 'Outside channel'
        }
      ]
    },
    {
      label: 'Percent change',
      options: [
        {
          value: 'percent_up',
          label: 'Moving up %'
        },
        {
          value: 'percent_down',
          label: 'Moving down %'
        }
      ]
    }
  ],
  daily_active_addresses: [
    { label: 'Daily Active Addresses', value: 'daily_active_addresses' }
  ],
  price_volume_difference: [
    { label: 'Price/volume difference', value: 'price_volume_difference' }
  ]
}

const ARGS = {
  price_volume_difference: ['threshold'],
  daily_active_addresses: ['percentThreshold', 'timeWindow'],
  price_percent_change: ['percentThreshold', 'timeWindow']
}

const defaultValues = {
  price_percent_change: {
    cooldown: '24h',
    percentThreshold: 5,
    target: { value: 'santiment', label: 'santiment' },
    timeWindow: 24,
    timeWindowUnit: { label: 'hours', value: 'h' },
    type: { value: 'price_percent_change', label: 'Moving up %' },
    operation: { value: 'percent_up', label: 'Moving up %' },
    metric: { label: 'Price', value: 'price' },
    isRepeating: true,
    channels: ['telegram']
  },
  daily_active_addresses: {
    cooldown: '24h',
    percentThreshold: 200,
    target: { value: 'santiment', label: 'santiment' },
    metric: {
      label: 'Daily Active Addresses',
      value: 'daily_active_addresses'
    },
    timeWindow: 2,
    timeWindowUnit: { label: 'days', value: 'd' },
    type: { label: 'Daily Active Addresses', value: 'daily_active_addresses' },
    isRepeating: true,
    channels: ['telegram']
  },
  price_volume_difference: {
    cooldown: '24h',
    threshold: 0.002,
    target: { value: 'santiment', label: 'santiment' },
    metric: {
      label: 'Price/volume difference',
      value: 'price_volume_difference'
    },
    type: {
      label: 'Price/volume difference',
      value: 'price_volume_difference'
    },
    isRepeating: true,
    channels: ['telegram']
  }
}

const DEFAULT_FORM_SETTINGS = defaultValues['price_percent_change']

const getTypeByMetric = metric => {
  if (metric.value === 'price') {
    return {
      value: 'price_percent_change',
      label: 'Moving up %'
    }
  } else {
    return TYPES[metric.value][0]
  }
}

const mapValuesToTriggerProps = values => ({
  cooldown: values.cooldown,
  settings: (() => {
    switch (values.metric.value) {
      case 'daily_active_addresses':
        return {
          percent_threshold: values.percentThreshold,
          target: { slug: values.target.value },
          time_window: values.timeWindow + values.timeWindowUnit.value,
          type: values.type.value
        }
      case 'price_volume_difference':
        return {
          target: { slug: values.target.value },
          threshold: values.threshold,
          type: values.type.value
        }
      default:
        return {
          percent_threshold: values.percentThreshold,
          target: { slug: values.target.value },
          time_window: values.timeWindow + values.timeWindowUnit.value,
          type: values.type.value
        }
    }
  })()
})

const validate = values => {
  let errors = {}

  if (
    values.type.value === 'daily_active_addresses' ||
    values.type.value === 'price_percent_change'
  ) {
    if (!values.percentThreshold) {
      errors.percentThreshold = 'Required'
    } else if (values.percentThreshold <= 0) {
      errors.percentThreshold = 'Must be more 0'
    }
    if (!values.timeWindow) {
      errors.timeWindow = 'Required'
    } else if (values.timeWindow <= 0) {
      errors.timeWindow = 'Must be more 0'
    }
  }
  if (values.type.value === 'price_volume_difference') {
    if (!values.threshold) {
      errors.threshold = 'Required'
    } else if (values.threshold < 0) {
      errors.threshold = 'Must be more 0'
    }
  }
  if (values.channels.length === 0) {
    errors.channels = 'You must setup notification channel'
  }
  return errors
}

const getMetricsByType = type => {
  switch (type.value) {
    case 'daily_active_addresses':
      return ['active_addresses', 'price']
    case 'price_volume_difference':
      return ['price', 'volume']
    default:
      return ['price']
  }
}

const DEFAULT_FORM_META_SETTINGS = {
  target: {
    isDisabled: false
  },
  type: {
    isDisabled: false
  },
  metric: {
    isDisabled: false
  }
}

const propTypes = {
  onSettingsChange: PropTypes.func.isRequired,
  isTelegramConnected: PropTypes.bool.isRequired
}

export const TriggerForm = ({
  onSettingsChange,
  getSignalBacktestingPoints,
  data: { allProjects = [] },
  isTelegramConnected = false,
  settings = DEFAULT_FORM_SETTINGS
  // metaFormSettings
}) => {
  // metaFormSettings = { ...DEFAULT_FORM_META_SETTINGS, ...metaFormSettings }
  // settings = {
  // ...settings,
  // target: metaFormSettings.target.value
  // ? metaFormSettings.target.value
  // : settings.target,
  // metric: metaFormSettings.metric.value
  // ? metaFormSettings.metric.value
  // : settings.metric,
  // type: metaFormSettings.type.value
  // ? metaFormSettings.type.value
  // : settings.type
  // }
  const [initialValues, setInitialValues] = useState(settings)

  // useEffect(() => {
  // getSignalBacktestingPoints(mapValuesToTriggerProps(initialValues))
  // }, [])

  return (
    <Formik
      initialValues={initialValues}
      isInitialValid
      enableReinitialize
      validate={validate}
      onSubmit={values => {
        onSettingsChange(values)
      }}
    >
      {({
        values,
        errors,
        isSubmitting,
        handleChange,
        handleBlur,
        setFieldValue,
        isValid,
        validateForm
      }) => (
        <Form>
          <FormikEffect
            onChange={(current, prev) => {
              console.table(current.values)
              if (current.values.type !== prev.values.type) {
                setInitialValues(defaultValues[current.values.type.value])
                validateForm()
                // window.setTimeout(() => {
                // getSignalBacktestingPoints(
                // mapValuesToTriggerProps(
                // defaultValues[current.values.type.value]
                // )
                // )
                // }, 0)
                return
              }
              if (!isEqual(current.values, prev.values)) {
                const lastErrors = validate(current.values)
                const isError = Object.keys(current.values).reduce(
                  (acc, val) => {
                    if (lastErrors.hasOwnProperty(val)) {
                      acc = true
                    }
                    return acc
                  },
                  false
                )

                //! !current.values.target &&
                //! isError &&
                // getSignalBacktestingPoints(mapValuesToTriggerProps(values))
              }
            }}
          />
          <div className={styles.Trigger}>
            <Panel header='Trigger #1' className={styles.TriggerPanel}>
              <div className={styles.row}>
                <div className={styles.Field}>
                  <label>Asset</label>
                  <FormikSelect
                    name='target'
                    // isDisabled={metaFormSettings.target.isDisabled}
                    // defaultValue={metaFormSettings.target.value}
                    placeholder='Pick an asset'
                    options={allProjects.map(asset => ({
                      label: asset.slug,
                      value: asset.slug
                    }))}
                  />
                </div>
              </div>

              <div className={styles.row}>
                <label>Metrics</label>
              </div>
              <div className={styles.row}>
                <div className={styles.Field}>
                  <FormikSelect
                    name='metric'
                    isClearable={false}
                    // isDisabled={metaFormSettings.metric.isDisabled}
                    // defaultValue={metaFormSettings.metric.value}
                    isSearchable
                    placeholder='Choose a metric'
                    options={METRICS}
                    onChange={metric => {
                      metric.value !== values.metric.value &&
                        setFieldValue('type', getTypeByMetric(metric))
                    }}
                  />
                </div>
                {TYPES[(values.metric || {}).value] &&
                  TYPES[(values.metric || {}).value].length > 1 && (
                  <div className={styles.Field}>
                    <FormikSelect
                      name='type'
                      isClearable={false}
                      isSearchable
                      // isDisabled={metaFormSettings.type.isDisabled}
                      // defaultValue={metaFormSettings.type.value}
                      placeholder='Choose a type'
                      options={TYPES[values.metric.value]}
                      isOptionDisabled={option => !option.value}
                    />
                  </div>
                )}
              </div>

              <div className={styles.row}>
                {ARGS[values.type.value].includes('percentThreshold') && (
                  <div className={styles.Field}>
                    <label>Percentage change</label>
                    <FormikInput
                      name='percentThreshold'
                      type='number'
                      placeholder='Setup the percentage change'
                    />
                  </div>
                )}
                {ARGS[values.type.value].includes('threshold') && (
                  <div className={styles.Field}>
                    <label>Threshold</label>
                    <FormikInput
                      name='threshold'
                      step={0.001}
                      type='number'
                      placeholder='Setup the threshold'
                    />
                  </div>
                )}
                {ARGS[values.type.value].includes('timeWindow') && (
                  <div className={styles.Field}>
                    <label>Time Window</label>
                    <div className={styles.timeWindow}>
                      <FormikInput
                        name='timeWindow'
                        type='number'
                        className={styles.timeWindowInput}
                        placeholder='Setup the time window'
                      />
                      <FormikSelect
                        name='timeWindowUnit'
                        className={styles.timeWindowUnit}
                        clearable={false}
                        placeholder='Choose unit'
                        options={[
                          { value: 'h', label: 'hours' },
                          { value: 'd', label: 'days' }
                        ]}
                      />
                    </div>
                  </div>
                )}
              </div>
              <div className={styles.row}>
                <div className={styles.Field}>
                  <label>Message Frequency</label>
                  <div>Once per</div>
                  <div className={styles.Field}>
                    <FormikSelector name='cooldown' options={['1h', '24h']} />
                  </div>
                </div>
              </div>
              <div className={styles.row}>
                <div className={styles.Field}>
                  <div className={styles.isRepeating}>
                    <FormikToggle name='isRepeating' />
                    <span>
                      {values.isRepeating
                        ? 'Task never ends'
                        : 'Task fires only once'}
                    </span>
                  </div>
                </div>
              </div>
              <label>Notify me via</label>
              <div className={styles.row}>
                <div className={styles.Field}>
                  <FormikCheckboxes
                    name='channels'
                    disabledIndexes={'email'}
                    options={['email', 'telegram']}
                    styles={{ marginRight: 15 }}
                  />
                </div>
                {!isTelegramConnected && (
                  <Button
                    className={styles.connectLink}
                    variant='ghost'
                    as={Link}
                    to='/account'
                  >
                    Telegram<span className={styles.connectLink}>Connect</span>
                  </Button>
                )}
              </div>

              {errors.channels && (
                <div className={cx(styles.row, styles.messages)}>
                  <Message variant='warn'>{errors.channels}</Message>
                </div>
              )}

              {
                // <label>Visual Backtesting</label>
                // <SignalPreview
                // target={values.target.value}
                // initialMetrics={getMetricsByType(values.type)}
                // type={values.type}
                /// >
              }
            </Panel>
          </div>
          <div className={styles.controls}>
            <Button
              type='submit'
              disabled={!isValid || isSubmitting}
              isActive={isValid && !isSubmitting}
              variant={'fill'}
              accent='positive'
            >
              Continue
            </Button>
          </div>
        </Form>
      )}
    </Formik>
  )
}

TriggerForm.propTypes = propTypes

const mapStateToProps = state => ({
  isTelegramConnected: selectIsTelegramConnected(state)
})

const mapDispatchToProps = dispatch => ({
  getSignalBacktestingPoints: payload => {
    dispatch(fetchHistorySignalPoints(payload))
  }
})

const enhance = compose(
  connect(
    mapStateToProps,
    mapDispatchToProps
  ),
  graphql(allProjectsForSearchGQL)
)

export default enhance(TriggerForm)
