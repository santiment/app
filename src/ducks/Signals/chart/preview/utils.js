export const mapWithTimeseries = (items) =>
  items.map((item) => ({
    ...item,
    datetime: +new Date(item.datetime),
    index: item.datetime,
  }))

export const mapWithMidnightTime = (date) => {
  const datetime = new Date(date)
  datetime.setUTCHours(0, 0, 0, 0)
  return +new Date(datetime)
}

export const mapWithTimeseriesAndYCoord = (
  triggered,
  triggersBy,
  timeseries,
  toDayConversion = true,
) => {
  const { key, dataKey = key, historicalTriggersDataKey: mappingKey = dataKey } = triggersBy

  const mapped = triggered.map((point) => {
    const date = toDayConversion ? mapWithMidnightTime(point.datetime) : +new Date(point.datetime)
    const item = timeseries.find(({ datetime }) => datetime === date)

    const fromTimeseries = item ? item[dataKey] || item[key] : undefined

    const yCoord = fromTimeseries || point[mappingKey] || point['current']
    return { date, yCoord, ...point }
  })

  return mapped
}

export const cleanByDatakeys = (timeseries, dataKey) => {
  return timeseries.filter((item) => item[dataKey] !== undefined)
}

export const makeSameRange = (points, base) => {
  if (!base[0]) {
    return points
  }

  const date = base[0].datetime
  return points.filter(({ datetime }) => +new Date(datetime) > date)
}

export const mapToRequestedMetrics = (metrics, settings) =>
  metrics.map(({ key, alias: name = key, reqMeta, fetch }) => ({
    key,
    name,
    fetch,
    ...settings,
    ...reqMeta,
  }))

export const getAvailableCooldown = (baseCooldown) => {
  if (baseCooldown && (baseCooldown.indexOf('d') !== -1 || baseCooldown.indexOf('w') !== -1)) {
    return '1d'
  }

  return baseCooldown && baseCooldown.indexOf('m') !== -1 ? '1h' : baseCooldown
}
