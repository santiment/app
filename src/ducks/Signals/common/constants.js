export const ALERT_ROUTES = {
  ALERTS: '/alerts',
  ALERT: `/alerts/:id`,
}
