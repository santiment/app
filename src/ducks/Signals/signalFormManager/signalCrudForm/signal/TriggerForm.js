import React, { useState, useEffect, Fragment, useCallback } from 'react'
import PropTypes from 'prop-types'
import cx from 'classnames'
import { Formik, Form } from 'formik'
import isEqual from 'lodash.isequal'
import FormikEffect from '../../../../../components/formik-santiment-ui/FormikEffect'
import FormikLabel from '../../../../../components/formik-santiment-ui/FormikLabel'
import Button from '@santiment-network/ui/Button'
import RadioBtns from '@santiment-network/ui/RadioBtns'
import {
  METRIC_TO_TYPES,
  MIN_TITLE_LENGTH,
  MAX_TITLE_LENGTH,
  TRIGGER_FORM_STEPS,
  COMMON_PROPS_FOR_METRIC,
} from '../../../utils/constants'
import {
  couldShowChart,
  mapFormPropsToTrigger,
  validateTriggerForm,
  getDefaultFormValues,
  titleMetricValuesHeader,
  getNewTitle,
  getNewDescription,
} from '../../../utils/utils'
import { TriggerFormMetricValues } from '../formParts/TriggerFormMetricValues'
import TriggerFormMetricTypes from '../formParts/metricTypes/TriggerFormMetricTypes'
import { TriggerFormFrequency } from '../formParts/TriggerFormFrequency'
import SignalPreview from '../../../chart/preview/SignalPreview'
import TriggerMetricTypesResolver from '../formParts/TriggerMetricTypesResolver'
import TriggerFormBlock, { TriggerFormBlockDivider } from '../formParts/block/TriggerFormBlock'
import FormikInput from '../../../../../components/formik-santiment-ui/FormikInput'
import TriggerFormChannels from '../formParts/channels/TriggerFormChannels'
import FormikCheckbox from '../../../../../components/formik-santiment-ui/FormikCheckbox'
import SignalFormDescription from '../formParts/description/SignalFormDescription'
import { useUserSettings } from '../../../../../stores/user/settings'
import styles from './TriggerForm.module.scss'

const getTitle = (formData, id, isShared) => {
  const isUpdate = id > 0 && !isShared
  const publicWord = formData.isPublic ? 'public' : 'private'
  if (isUpdate) {
    return `Update ${publicWord} alert`
  } else {
    return `Create ${publicWord} alert`
  }
}

const propTypes = {
  onSettingsChange: PropTypes.func.isRequired,
  settings: PropTypes.any,
  metaFormSettings: PropTypes.any,
  triggerMeta: PropTypes.any,
}

export const TriggerForm = ({
  id,
  onSettingsChange,
  settings = {},
  metaFormSettings,
  formChangedCallback,
  isShared,
  setTitle,
}) => {
  const isNew = !id
  const [initialValues, setInitialValues] = useState(settings)
  const [canCallFormChangCallback, setCanCallFormChanged] = useState(false)
  const [step, setStep] = useState(TRIGGER_FORM_STEPS.DESCRIPTION)

  useEffect(() => {
    if (!isNew && !isEqual(settings, initialValues)) {
      setInitialValues(settings)
    }
  }, [settings])

  useEffect(() => {
    const timeOutId = setTimeout(() => {
      !canCallFormChangCallback && setCanCallFormChanged(true)
    })

    return () => {
      clearTimeout(timeOutId)
      setCanCallFormChanged(false)
      formChangedCallback(false)
    }
  }, [])

  useEffect(() => {
    setTitle && setTitle(getTitle(initialValues, id, isShared))
  }, [initialValues.isPublic])

  const validateAndSetStep = useCallback(
    (newStep) => {
      if (isNew) {
        newStep > step && setStep(newStep)
      }
    },
    [isNew, step, setStep],
  )

  const {
    settings: { isTelegramAllowAlerts, isEmailAllowAlerts },
  } = useUserSettings()

  return (
    <Formik
      initialValues={initialValues}
      isInitialValid={!isNew}
      enableReinitialize
      validate={validateTriggerForm}
      onSubmit={(values) => {
        onSettingsChange(values)
      }}
    >
      {({
        values,
        errors,
        isSubmitting,
        handleChange,
        handleBlur,
        setFieldValue,
        isValid,
        validateForm,
      }) => {
        const {
          metric,
          type = {},
          target,
          targetWatchlist,
          frequencyType,
          frequencyTimeType,
          isRepeating,
          ethAddress,
          isPublic,
          description,
          channels = [],
        } = values

        const mappedTrigger = mapFormPropsToTrigger(values)

        const showChart = (target || targetWatchlist) && couldShowChart(mappedTrigger.settings)

        const typeSelectors = metric.key ? COMMON_PROPS_FOR_METRIC : METRIC_TO_TYPES[metric.value]

        const showTypes = typeSelectors && typeSelectors.length > 1

        const { dependencies: metricValueBlocks } = type

        const showValues = showTypes || showChart || metricValueBlocks

        if (!showValues && step === TRIGGER_FORM_STEPS.VALUES) {
          validateAndSetStep(TRIGGER_FORM_STEPS.DESCRIPTION)
        }

        const isValidForm = isValid || !errors || Object.keys(errors).length === 0

        const showDivider = showTypes || metricValueBlocks

        function toggleSignalPublic(isPublicTarget) {
          if (isPublicTarget !== isPublic) {
            setFieldValue('isPublic', isPublicTarget)
          }
        }

        return (
          <Form>
            <FormikEffect
              onChange={(current, prev) => {
                let { values: newValues } = current

                if (!isEqual(newValues, prev.values)) {
                  const changedMetric =
                    !prev.values.metric ||
                    newValues.metric.value !== prev.values.metric.value ||
                    newValues.type.value !== prev.values.type.value

                  if (changedMetric) {
                    setInitialValues(getDefaultFormValues(newValues, prev.values.metric))
                  }

                  validateForm()

                  if (isNew && !isShared) {
                    !newValues.titleChangedByUser && setFieldValue('title', getNewTitle(newValues))
                    !newValues.descriptionChangedByUser &&
                      setFieldValue('description', getNewDescription(newValues))
                  }

                  canCallFormChangCallback && formChangedCallback && formChangedCallback(true)
                }
              }}
            />

            <div className={styles.triggerFormItem}>
              {step >= TRIGGER_FORM_STEPS.METRICS && (
                <TriggerFormMetricTypes
                  metaFormSettings={metaFormSettings}
                  setFieldValue={setFieldValue}
                  metric={metric}
                  target={target}
                  trigger={mappedTrigger}
                />
              )}

              {step >= TRIGGER_FORM_STEPS.TYPES && (
                <TriggerMetricTypesResolver
                  address={ethAddress}
                  values={values}
                  metaFormSettings={metaFormSettings}
                  setFieldValue={setFieldValue}
                  isNewSignal={isNew}
                />
              )}

              {step >= TRIGGER_FORM_STEPS.VALUES && showValues && (
                <TriggerFormBlock
                  {...titleMetricValuesHeader(!!metricValueBlocks, values)}
                  className={styles.chainBlock}
                >
                  {showDivider && (
                    <TriggerFormMetricValues
                      typeSelectors={typeSelectors}
                      metaFormSettings={metaFormSettings}
                      blocks={metricValueBlocks}
                      values={values}
                      showTypes={showTypes}
                    />
                  )}

                  {showChart && (
                    <>
                      {showDivider && <TriggerFormBlockDivider />}
                      <div className={cx(styles.preview, showDivider && styles.previewWithDiviver)}>
                        <SignalPreview trigger={mappedTrigger} type={metric.value} />
                      </div>
                    </>
                  )}
                </TriggerFormBlock>
              )}

              {step >= TRIGGER_FORM_STEPS.DESCRIPTION && (
                <Fragment>
                  <TriggerFormBlock
                    titleLabel='More options'
                    showDescription={false}
                    enabledHide
                    show={!isTelegramAllowAlerts || !isEmailAllowAlerts}
                    className={styles.chainBlock}
                  >
                    <TriggerFormChannels
                      isNew={isNew}
                      channels={channels}
                      errors={errors}
                      setFieldValue={setFieldValue}
                    />

                    <TriggerFormBlockDivider />

                    <TriggerFormFrequency
                      disabled={!isRepeating}
                      metaFormSettings={metaFormSettings}
                      setFieldValue={setFieldValue}
                      frequencyType={frequencyType}
                      metric={type.metric}
                      frequencyTimeType={frequencyTimeType}
                    />

                    <div className={cx(styles.row, styles.rowTop)}>
                      <FormikCheckbox
                        className={styles.isRepeating}
                        name='isRepeating'
                        isActive={!isRepeating}
                        label='Disable after it triggers'
                        onClick={() => {
                          setFieldValue('isRepeating', !isRepeating)
                        }}
                      />
                    </div>

                    <TriggerFormBlockDivider />

                    <div className={styles.row}>
                      <div className={styles.Field}>
                        <FormikLabel text='Signal visibility' />
                        <div className={styles.triggerToggleBlock}>
                          <RadioBtns
                            options={['Public', 'Private']}
                            labelOnRight
                            defaultSelectedIndex={isPublic ? 'Public' : 'Private'}
                            passSelectionIndexToItem
                            onSelect={(value) => {
                              toggleSignalPublic(value === 'Public')
                            }}
                            labelClassName={styles.checkboxLabel}
                          />
                        </div>
                      </div>
                    </div>
                  </TriggerFormBlock>

                  <div className={cx(styles.row, styles.descriptionBlock)}>
                    <div className={cx(styles.Field, styles.fieldFilled)}>
                      <FormikLabel text='Name of the alert' />
                      <FormikInput
                        name='title'
                        type='text'
                        minLength={MIN_TITLE_LENGTH}
                        maxLength={MAX_TITLE_LENGTH}
                        placeholder='Name of the alert'
                        onChange={() => setFieldValue('titleChangedByUser', true)}
                      />
                    </div>
                  </div>

                  <div className={styles.row}>
                    <SignalFormDescription
                      setFieldValue={setFieldValue}
                      description={description}
                    />
                  </div>
                </Fragment>
              )}

              <Button
                type='submit'
                disabled={!isValidForm}
                isActive={isValidForm}
                variant={'fill'}
                accent='positive'
                className={styles.submitButton}
              >
                {id && !isShared ? 'Update alert' : 'Create alert'}
              </Button>
            </div>
          </Form>
        )
      }}
    </Formik>
  )
}

TriggerForm.propTypes = propTypes

export default TriggerForm
