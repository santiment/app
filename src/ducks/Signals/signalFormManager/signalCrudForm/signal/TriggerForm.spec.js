/* eslint-env jest */
import React from 'react'
import { shallow } from 'enzyme'
import toJson from 'enzyme-to-json'
import { TriggerForm } from './TriggerForm'
import { ETH_WALLET_METRIC } from '../../../utils/constants'
import { ApolloProvider } from 'react-apollo'
import { client } from '../../../../../apollo'

const allProjects = [
  {
    slug: 'santiment',
  },
]

describe('TriggerForm', () => {
  it('smoke', () => {
    const mockCb = jest.fn()
    const wrapper = shallow(
      <ApolloProvider client={client}>
        <TriggerForm
          data={allProjects}
          canRedirect={true}
          isTelegramConnected={true}
          onSettingsChange={mockCb}
          getSignalBacktestingPoints={() => {}}
        />
      </ApolloProvider>,
    )
    expect(toJson(wrapper)).toMatchSnapshot()
  })
})

describe('TriggerForm EthWallet', () => {
  it('smoke', () => {
    const mockCb = jest.fn()

    const asset = 'nuga'
    const address = 'some_custom_address'

    const wrapper = shallow(
      <ApolloProvider client={client}>
        <TriggerForm
          data={allProjects}
          canRedirect={true}
          isTelegramConnected={true}
          onSettingsChange={mockCb}
          getSignalBacktestingPoints={() => {}}
          metaFormSettings={{
            target: {
              value: {
                value: asset,
                label: asset,
              },
            },
            metric: {
              value: { ...ETH_WALLET_METRIC },
            },
            type: {
              value: { ...ETH_WALLET_METRIC },
            },
            ethAddress: address,
          }}
        />
      </ApolloProvider>,
    )
    expect(toJson(wrapper)).toMatchSnapshot()
  })
})
