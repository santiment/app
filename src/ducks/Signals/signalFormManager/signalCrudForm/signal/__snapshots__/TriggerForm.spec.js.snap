// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`TriggerForm smoke 1`] = `
<Formik
  enableReinitialize={true}
  initialValues={
    Object {
      "address": undefined,
      "channels": Array [
        "telegram",
      ],
      "frequencyTimeType": Object {
        "label": "Hours",
        "value": "h",
      },
      "frequencyTimeValue": Object {
        "label": 1,
        "value": 1,
      },
      "frequencyType": Object {
        "label": "Once Per",
        "value": "onceper",
      },
      "isRepeating": true,
      "metric": Object {
        "label": "Price",
        "value": "price",
      },
      "percentThreshold": 5,
      "signalType": Object {
        "label": "Assets",
        "value": "assets",
      },
      "target": Object {
        "label": "santiment",
        "value": "santiment",
      },
      "timeWindow": 24,
      "timeWindowUnit": Object {
        "label": "hours",
        "value": "h",
      },
      "type": Object {
        "label": "Moving up %",
        "type": "percent_up",
        "value": "price_percent_change",
      },
    }
  }
  isInitialValid={true}
  onSubmit={[Function]}
  validate={[Function]}
  validateOnBlur={true}
  validateOnChange={true}
>
  <Component />
</Formik>
`;
